kernel-install: fix invocation as installkernel, add tests, tweak documentation

keszybz opened the pull request.

keszybz commented:
Oops, I did some last-minute changes and screwed up the syntax. Should be better now.

DaanDeMeyer commented on man/kernel-install.xml:
```diff
@@ -173,77 +173,91 @@
+      <varname>$KERNEL_INSTALL_MACHINE_ID</varname> are often set to the same value, the latter is guaranteed
+      to be a valid 32 character ID in lowercase hexadecimals while the former can be any short string. The
+      entry token to use is read from <filename>/etc/kernel/entry-token</filename>, if it exists. Otherwise a
+      few possible candidates below the <varname>$BOOT</varname> are checked for Boot Loader Specification
```
$BOOT directory?

DaanDeMeyer commented on man/kernel-install.xml:
```diff
@@ -173,77 +173,91 @@
+      be overridden by setting <varname>$BOOT_ROOT</varname> (see below).</para>
+
+      <para><varname>$KERNEL_INSTALL_LAYOUT=bls|other|...</varname> is set for the plugins to specify the
+      installation layout.  Defaults to <option>bls</option> if
```
Spurious whitespace

keszybz replied:
It's an emacs thing. Your comment prompted me to finally take care of this. It turns out that it's trivial:
```
(setq sentence-end-double-space nil)
```
@poettering please also apply this to your config.

DaanDeMeyer commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -104,10 +104,11 @@ if ! [ -d "$ENTRY_DIR_ABS" ]; then
     echo "Error: could not copy '$KERNEL_IMAGE' to '$ENTRY_DIR_ABS/linux'." >&2
     exit 1
 }
+chown root.root "$ENTRY_DIR_ABS/linux" || :
```
I'm not sure about suddenly silently failing when we can't change ownership. Should we do this at all?

keszybz replied:
It's not silent. `chown` prints an error.  I think it's fine to do this. In practice this only matters when running as non-root, i.e. in some test scenario.

medhefgo replied:
In practice, these should almost always be written to fat, no? chown doesn't work on fat, so it'll always complain.

DaanDeMeyer reviewed (changes requested):
I'd split the `test -f` changes into its own commit separate from the CONF_ROOT stuff. Aside from that, a few nits but overalll LGTM. Great to have a test for kernel-install.

keszybz commented:
Updated. I took the plunge and added `set -e` to `kernel-install`. Together with the change to use `test -f` this means we'll fail if we cannot read a config file. Overall, I think this is good, we should never ignore errors in something as important as installation of boot loader entires. I just hope this doesn't result in some strange regression.

medhefgo commented on src/kernel-install/test-kernel-install.sh:
```diff
@@ -0,0 +1,83 @@
+"$kernel_install" inspect
+
+"$kernel_install" -v remove 1.1.1
+( ! test -f "$entry" )
```
Why the `()`?  Why not use `test ! -f "$entry"` these?

keszybz replied:
Yeah, no reason. Unfortunately there's one `! grep`, so this `()` is still present in one place.

keszybz replied:
It would… Question of style. I prefer the form with the explicit invert.

medhefgo replied:
It doesn't require silencing a linter…

poettering commented on src/kernel-install/kernel-install.in:
```diff
@@ -83,7 +83,7 @@ if [ "${0##*/}" = "installkernel" ]; then
     #   /sbin/installkernel <version> <vmlinuz> <map> <installation-dir>
     # We ignore the last two arguments.
-    set -- "$1"
+    set -- "$1" "$2"
```
hmm, to be more accurate, shouldn't this move further up, before we look for --help/-h/--version/--verbose/-v?

I mean those switches are not defined for "/sbin/installkernel" afaik, and one could argue that if passed anyway they should be understood as version parameter?

btw, should we refuse if less than 2 args are specified? simply to catch cases if people invoke the script under that name directly to see what it does?

keszybz replied:
> hmm, to be more accurate, shouldn't this move further up, before we look for --help/-h/--version/--verbose/-v?
>
> I mean those switches are not defined for "/sbin/installkernel" afaik, and one could argue that if passed anyway they should be understood as version parameter?

I think it's nicer to allow `--version`, `--verbose`, and `--help` to work with `installkernel`. The original script didn't support that, but we can. If somebody names their kernel `--help`, that's not our problem.
 
> btw, should we refuse if less than 2 args are specified? simply to catch cases if people invoke the script under that name directly to see what it does?

We should. I added a check now.

poettering commented on man/kernel-install.xml:
```diff
@@ -173,77 +174,91 @@
-    <filename>install.conf</filename> upon initial installation. The following values are currently
-    understood:</para>
+    <refsect2>
+      <title>Environment variables exported for plugins</title>
```
isn't it "export to", not "export for"?

keszybz replied:
I think both are OK. "Exported to" implies that something is moved across the boundary. "Exported for" implies that kernel-install does the operation of exporting for the benefit of the plugins.

poettering commented on man/kernel-install.xml:
```diff
@@ -173,77 +174,91 @@
+    </refsect2>
+
+    <refsect2>
+      <title>Environment variables used by <command>kernel-install</command></title>
```
used? i mean, in a way it also "uses" the variables it passes down to the plugins. Maybe say "accepted"?

keszybz replied:
I think we use "understood" in similar contexts. Maybe that'd be better.

poettering commented on src/kernel-install/kernel-install.in:
```diff
@@ -108,15 +108,17 @@ initrd_generator=
 _MACHINE_ID_SAVED="$MACHINE_ID"
 _BOOT_ROOT_SAVED="$BOOT_ROOT"
 
-if [ -f "/etc/kernel/install.conf" ]; then
```
wouldn't it be easier to just always set KERNEL_INSTALL_CONF_ROOT if not set already, and then always use that var? i.e. add:

```sh
if [ -z "$KERNEL_INSTALL_CONF_ROOT" ]; then
        KERNEL_INSTALL_CONF_ROOT=/etc/kernel
fi
```
That should make things easier when referencing the dir, because we can then always just use the var

keszybz replied:
The variable is used for multiple files, and they need different logic. E.g. for `install.conf` we read /etc/ and /usr/lib/, but for `tries`, only `/etc/`. So I think it's better to keep it as is.

poettering commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -121,10 +122,11 @@ for initrd in "$@" "${KERNEL_INSTALL_STAGING_AREA}"/initrd*; do
         echo "Error: could not copy '$initrd' to '$ENTRY_DIR_ABS/$initrd_basename'." >&2
         exit 1
     }
+    chown root.root "$ENTRY_DIR_ABS/$initrd_basename" || :
```
the C programmer in mind kinda hates the fact that this means the file cannot be created atomically with the right ownership anymore

keszybz commented on src/kernel-install/kernel-install.in:
```diff
@@ -312,11 +312,15 @@ MAKE_ENTRY_DIR_ABS=$?
-        "/usr/lib/kernel/install.d"
-)"
+if [ -n "$KERNEL_INSTALL_PLUGINS" ]; then
+    PLUGINS="$KERNEL_INSTALL_PLUGINS"
```
I was mostly conforming to the style of surrounding code… I don't have a strong opinion on this.

keszybz replied:
Updated.

keszybz commented:
Updated. I addressed all comments, except as mentioned above.

> `[ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && echo …` should really be replaced with a `log_verbose` helper if you ask me…

Sure. But let's do that separate, this PR is long enough.

poettering commented on man/kernel-install.xml:
```diff
@@ -255,6 +255,11 @@
 
+      <para><varname>$KERNEL_INSTALL_PLUGINS</varname> can be set to override the list of plugins executed by
+      <command>kernel-install</command>. The argument is a newline-separated list of paths.
+      <literal>KERNEL_INSTALL_PLUGINS=:</literal> may be used to prevent any plugins from running.
```
hmm, newline separated? that's quite unusual. Sounds like unnecessarily hard to use? why not just use ":" or some more commonly used seperator in shell scripts?

It's not that we really should care about people who are crazy enough to encode ":" in their plugin name.

keszybz replied:
Updated with "whitespace-seperated list".

poettering commented:
lgtm. no the superduper shell guru myself, but this looks all ok to me now, and i guess this went through its number of reviews already.
